frontend/device/gesturedetector.lua

```diff
@@ -461,22 +468,9 @@ function GestureDetector:handleNonTap(tev)
         -- switched from other state, probably from initialState
         -- we return nil in this case
         self.states[slot] = self.tapState
-        if self._has_real_clock_time_ev_time == nil then
-            if tev.timev.sec < TimeVal:now().sec - 600 then
-                -- ev.timev is probably the uptime since device boot
-                -- (which might pause on suspend) that we can't use
-                -- with setTimeout(): we'll use TimeVal:now()
-                self._has_real_clock_time_ev_time = false
-                logger.info("event times are not real clock time: some adjustments will be made")
-            else
-                -- assume they are real clock time
-                self._has_real_clock_time_ev_time = true
-                logger.info("event times are real clock time: no adjustment needed")
-            end
-        end
         logger.dbg("set up hold timer")
-        local ref_time = self._has_real_clock_time_ev_time and tev.timev or TimeVal:now()
-        local deadline = ref_time + TimeVal:new{
+        -- Ditto, start counting from *now*, no matter what the input event timestamp says...
+        local deadline = TimeVal:monotonic() + TimeVal:new{
```

Comment:
You're sure about all that ? (I'm not putting my mind again in all that, but thinking about 633d43b322 #6798, I'm wondering if you're not missing one of the 2 cases this was supposed to handle both correctly?)

Reply:
We actually care about computing deadlines based on the moment we *process* them, not whatever time the kernel said they came from.

Reply:
But wasn't there a thing with:
`self._has_real_clock_time_ev_time and tev.timev `
trusting and actually using the time from the event (which might be some time in the past, and not a "now") ?

Or is it just that part, and this disctinction is still used for other things around there? (sorry, too busy to look at the code right now)

Reply:
> We actually care about computing deadlines based on the moment we process them

I think this part of the code was actually meant to compute a deadline from the ev.time if we can trust it - and fallback to now() only when we don't trust it.

Reply:
My point is that trusting it doesn't make much sense to me right now, which is what makes me think that the point is moot ^^.

Reply:
> My point is that trusting it doesn't make much sense to me right now

I think it made sense _at the time_ :)

We get events, that may be queued because we're doing some other processing - then, at the time we can process them, some time has elapsed. Using "now" might increase the interval since the time the user made that event.
I think this part of the code handles double tap, tap intervals, etc... and using now() was an issue. Don't want to re-dig into all that. Some more context around https://github.com/koreader/koreader/pull/6798#issuecomment-711154547 maybe.

Reply:
This part is only dealing with the actual event loop deadline; gesture detection itself still trusts the event's timestamps, as it should. (e.g., we compare apples to apples (the event loop), and oranges to oranges (input events themsevles) ;).

Reply:
I'm not convinced :/
These 2 parts set up a delay (a max interval between current event A and a futur event B) after which we change some internal state (so we process that futur event B one way, or another, depending on its delay from event A).
Previously, we used the real event time (which you can make monotonic() if you want) which might be a few 100ms in the past (because we've been busy and our UI loop is catching up, unqueuing past events) as the start of the delay - AND ONLY now() when we knew this ev.time was bad.
So, if:
ev.time = 13h00m00s
now (monotonic if you want) = 13h00m03s
and we have a hold_interval of 5s, we want the timeout to be triggered at 13h00m05s, not at 13h00m08s.

Reply:
Sure, but is there actually a practical example of when this might matter?

Reply:
So, it's for me to look at the hyper long #6798, where the thing you quickly removed was painfully added, to find out why we ended up adding it ? :/ 

Looks like I thought like you at https://github.com/koreader/koreader/pull/6798#issuecomment-711139118 : 
> So, it's just when handling double tap and hold/long-press that we would have a possibly shift between the event timestamp and the now() used for the timeout.
When everything is not busy, these should be quite similar and this change should have no impact.
And when we long-press or do double tap, things shouldn't be busy (when double tap, the first tap does not trigger anything so no reason to get busy).

I let you read @ezdiy and your comments after it, that you'll probably understand better than I do.
I guess I just followed @ezdiy request to not drop event times, and to not change too much of how all that works as it was working fine.

My only comment after that, possibly telling what this first naive change (always using now()) might have caused, although I'm not certain I wrote it about that, is:

> Anyway, I still have some issues with swipe sometimes being detected as multiswipe... (small diagonal swipe to full refresh, working sometimes, sometimes not...)

--

> Sure, 

Glad you understand the issue.
Well, when things are accurate and can be kept accurate, and they can cause a delay and we're playing with 50-500 ms delays, I don't see why you would want to make them less accurate.
And furthermore, with this, you'll be adding some expensive os.realtime() call, while you could just benefit of the one computed at the time the event was made - that should appeal you :)

Reply:
I'll triple-check, but I'm of the mind that trusting the event's timestamp would be a red herring here ;).

And, so far, nothing's broke on a device with events using a REALTIME clock source.

Reply:
AFAICT, on the extremely unlikely case we're significantly behind in reading the input events, it doesn't really matter if the setTimeout deadlines are further than they would be if the event timestamps were honored:

* The gesture detection itself will use the event timestamps, as it should be, so the detection itself is always doing the right thing.
* The only difference is that the Gesture Event will be sent "later", which is potentially a *good* thing:
   Suppose we were trusting the events to compute the deadline: that means we can essentially setup a deadline in the *past*, which means it will possibly expire *before* we've consumed the following events that could make it true.

Now, hitting a stall will unwind the event queue in a consistent manner, and Gestures will be detected as if stuff was not late, instead of potentially dropping early gestures because we caught the events "late".
At worst we end up with a bunch of extra deadlines that'll all expire later than they should and do nothing, which is harmless.

It's somewhat hard to test in practice because nothing much really uses this codepath except the tap -> hold switch, but I really can't make it do anything obviously wrong, even with a slow-ass rotated PDF (on a device with realtime events).

Reply:
Now, if you remember something that's easy to break to stress-test this, I'm all ears ;).

Reply:
Even on a design level, setTimeout is operating on a completely different timescale: it's concerned with the event loop polling, not the events themselves. As such, make it use the same timescale as the UI just makes more sense that trying to shoehorn potentially unreliable input events timestamps in there.

Reply:
> 
>     * The only difference is that the Gesture Event will be sent "later", which is potentially a _good_ thing:
>       Suppose we were trusting the events to compute the deadline: that means we could essentially setup a deadline in the _past_, which means it will possibly expire _before_ we've consumed the following events that could make it true and actually detect the gesture it ought to.

Actually, not even that will differ: the `waitEvent` only checks the deadlines when the event queue has been completely drained (i.e., when the waitForEvent select times out).

Which means that we'll just finalize the Gesture detection "late" (relative to how "late" we were to begin with).

Reply:
There's also the fact that the first deadline hit that detects a gesture "wins", and clears *all* the deadlines.
Being "late" doesn't change that fact, though, so, basically still business as usual, it'll just clear more deadlines than usual.

Reply:
And since #6798 was concerned abound bounces: that works just fine (and doesn't even rely on setTimeout in the first place anyway :D).

Reply:
As for why gloves were necessary before this PR, as in #6798: that's because the event loop was using a timescale that it naively thought would match the input events', so precautions had to be taken in case mismatches were detected.

This PR takes a different approach: it enforces a monotonic timescale in the UI/event loop, making scheduling reliable; and treats the input events timestamps as sacred, but potentially unreliable and in a different timescale: as such, anything UI/scheduling related follows the same timescale (monotonic), and input events are handled in their own little world by themselves, never even attempting to mix the two time scale.

Reply:
Thanks for digging deeper into all that. I'm not following everything, but I trust your feeling it is ok/better.

> that means we could essentially setup a deadline in the past

Somehow hoping this might solve my issue of touch getting locked - and me being in a hold state in some cases when I recover.
So ok, let's go with it and see. May be for after 2021.03 ?

Reply:
Happy too, writing it down helps me understand what happens and whether the design decision made sense ;).

I would assume the "sticky" state on resume was because the gesture detection code did not deal with large time deltas very well. That *should* be fixed, yeah ;).

https://github.com/koreader/koreader/issues/7340#issuecomment-799710372 is what spurred me on, but since it'll only "crash" w/ debug enabled, I'm fine with postponing this for 2021.04 (assuming it lands this WE ;)) ;).

Reply:
Btw, taking a step further away: are setTimeout() the only solution to this double tap / long-press interval ? It seems really complicated when I had to understand it some months ago...
No simpler way to handle this, like just storing timestamp in variables and considering then when next event is coming, if they have expired or not, ... ? (Not asking you to revamp the whole thing, just thinking if there's room there to do it better later.) (although if it works... don't fix it :)

Reply:
Yeah, preaching to the choir here ;p.

It indeed feels like an extremely roundabout way to detect a hold, but this is an old and somewhat key piece of code, and I'm not really up to rewriting gesturedetector ;p.